data_safe_haven/administration/users/active_directory_users.py

```diff
@@ -83,9 +90,10 @@ def list(self, sre_name: str | None = None) -> Sequence[ResearchUser]:  # noqa:
             self.vm_name,
         )
         users = []
+        expected_columns = 6
```

Comment:
How about replacing lines 93-112 with this, which will check that there are exactly 6 tokens?

```suggestion
        for line in output.split("\n"):
            try:
                sam_account_name, given_name, surname, phone_number, email_address, user_principal_name = line.split(";")
                users.append(
                    ResearchUser(
                        email_address=email_address,
                        given_name=given_name,
                        phone_number=phone_number,
                        sam_account_name=sam_account_name,
                        surname=surname,
                        user_principal_name=user_principal_name,
                    )
                )
            except ValueError as exc:
                msg = "Unexpected number of fields returned for Active Directory user."
                raise DataSafeHavenActiveDirectoryError(msg)
```

Reply:
That's fine I think (assuming it works).

The original had `if len(tokens) >= 6:` but here the list must have exactly 6 items. Will it ever be more than 6?

I would rather use an if/else block than try/except. I think it is clearer and _usually_ more performant.

Reply:
Here's the script https://github.com/alan-turing-institute/data-safe-haven/blob/magic_number/data_safe_haven/resources/active_directory/list_users.ps1. We have control over what it returns, which is currently `"$($User.SamAccountName);$($User.GivenName);$($User.Surname);$($User.TelephoneNumber);$($User.Mail);$($User.UserPrincipalName)"`

The reason for `try/except` here is that `a, b, c = <sequence>` will raise a `ValueError` if there are not exactly 3 elements in `<sequence>`, so we need to catch that.

Reply:
The script does seem to check that the required fields exist (at least some of them, maybe the others are mandatory?). Hard to say with confidence but it looks like it would always write six items separated by `;`, although some of them may be empty.

Probably still worth dealing with problems here though as the powershell script is external, could be out of sync.

> The reason for try/except here is that a, b, c = <sequence> will raise a ValueError if there are not exactly 3 elements in <sequence>, so we need to catch that.

Sure, but you could also do something like

```Python
if len(sequence) == 6:
    ...
else:
    raise ....
```

Where the thing you testing for is clearer.

Reply:
Fixed in c8e9b7ab